### PR TITLE
Refactor Deepseek v4 to use Attention Gym's Compressed Sparse Attention

### PR DESCRIPTION
**Pull request by floatingtrees**
Compressed Sparse Attention (CSA) is a data dependent sparse attention mechanism. It does not work well with Flex Attention; Flex Attention uses a block sparsity mask to determine which tiles to compute, and it recomputes the block mask every time the mask changes. Since CSA is data dependent, this means Flex Attention recomputes the block mask on almost every forward pass. 

For the debug model, the Attention Gym operator is around 3x faster at 1024 sequence length and 11x faster at 2048 sequence length. For an end to end run of the debug model, model performance rises from 36 TFLOPS/GPU to 105 TFLOPS/GPU. 

Tests were performed on 2xGB200. 

This PR currently contains:
1. A refactor to `torchtitan/models/deepseek_v4/attention.py` that calls the Attention Gym operator instead of Flex Attention. 
2. Correctness tests.

### CODE REVIEW

File: torchtitan/models/deepseek_v4/attention.py
Comment (tianyu-l) [inline torchtitan/models/deepseek_v4/attention.py:283]: Is this the commonly seen sliding window attention? If so, why do we need to build ad hoc mask per forward, instead of only once per iteration?

Comment (tianyu-l) [inline torchtitan/models/deepseek_v4/attention.py]: This is still inefficient, right? Is there a plan for it?

File: torchtitan/models/deepseek_v4/attention.py [hunk 1]
```diff
@@ -322,6 +309,20 @@ def forward(  # pyrefly: ignore[bad-param-name-override]
 
 
 class CompressedSparseAttention(DSV4FlexAttention):
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v4/attention.py:327]: could you show some perf comparison?

File: torchtitan/models/deepseek_v4/attention.py
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py]: Yes, there's a PR in Attention Gym that adds a fused kernel for the indexer here, but it hasn't been merged yet. The current implementation is a workaround for now.

Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py:283]: Thanks for pointing this out, I'll add a cache for the mask.

File: torchtitan/models/deepseek_v4/attention.py [hunk 2]
```diff
@@ -222,51 +262,14 @@ def _forward_impl(
...
-                        bsz=1, seqlen=seqlen, n_cmp=n_cmp, device=q.device
-                    )
-                )
-            sink_indices = torch.full(
-                (1, seqlen, 1), sink_idx, dtype=torch.int64, device=q.device
-            )
-            selected_indices.append(sink_indices)
-            selected_indices = torch.cat(selected_indices, dim=-1)
+        block_mask = attention_masks
+        if block_mask is None:
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v4/attention.py:266]: Should never be `None` after this change?

**Review by tianyu-l — APPROVED** [review]
one remaining comment

File: torchtitan/models/deepseek_v4/attention.py (hunk 2 above)
Comment (drisspg) [reply torchtitan/models/deepseek_v4/attention.py:266]: +1

File: torchtitan/models/deepseek_v4/attention.py [hunk 3]
```diff
@@ -338,16 +355,73 @@ def forward(  # pyrefly: ignore[bad-param-name-override]
...
+        if not math.isclose(self.softmax_scale, expected_scale, rel_tol=1e-6):
+            raise ValueError(
+                "CompressedSparseAttention requires softmax_scale == "
+                f"head_dim**-0.5 ({expected_scale}), got {self.softmax_scale}; "
+                "selected_attention does not support a custom scale."
+            )
+
+        seqlen = q.size(0)
+
+        with spmd.no_typecheck():
```
Comment (drisspg) [inline torchtitan/models/deepseek_v4/attention.py:386]: TODO: We should open a PR in Attention Gym to add a keyword-only `scale` argument to `selected_attention`,

File: torchtitan/models/deepseek_v4/attention.py [hunk 4]
```diff
@@ -338,16 +355,73 @@ def forward(  # pyrefly: ignore[bad-param-name-override]
...
-            idx_k=idx_k,
-            idx_w=idx_w,
-            attention_masks=attention_masks,
-        )
+        if attention_masks is not None:
+            raise ValueError(
+                "CompressedSparseAttention does not accept attention_masks; "
+                "top-k selection is computed internally."
+            )
+        if attn_sink is None:
```
Comment (drisspg) [inline torchtitan/models/deepseek_v4/attention.py:363]: ill take another look at the fa4 change

File: torchtitan/models/deepseek_v4/attention.py [hunk 5]
```diff
@@ -338,16 +355,73 @@ def forward(  # pyrefly: ignore[bad-param-name-override]
...
+                q_bhtk,
+                swa_k_b1tk,
+                cmp_k_b1tk,
+                cmp_topk,
+                attention_sink=attn_sink,
+                doc_ids=None,
+                sliding_window_size=self.window_size,
+                backend=backend,
+            )
+            return out_bhtk.squeeze(0).transpose(0, 1)
```
Comment (drisspg) [inline torchtitan/models/deepseek_v4/attention.py:425]: I thought our Triton implementation didn't support head_dim=512 for training, and the FA4 path didn't have sinks enabled yet? How do we handle the production configs here?

File: torchtitan/models/deepseek_v4/model.py [hunk 6]
```diff
@@ -175,8 +177,36 @@ def __init__(self, config: Config):
...
+        for block in blocks:
+            assert isinstance(block, DeepSeekV4TransformerBlock)
+            inner = block.attention.inner_attention
+            if not isinstance(inner, DSV4FlexAttention):
+                continue
+            key = dsv4_mask_key(inner.compress_ratio)
+            if key is None or key in masks:
+                continue
+            assert window_sizes.setdefault(key, inner.window_size) == inner.window_size
+            assert block_sizes.setdefault(key, inner.block_size) == inner.block_size
```
Comment (drisspg) [inline torchtitan/models/deepseek_v4/model.py:201]: doesn't `key in masks` skip subsequent layers before these window/block-size assertions run? lets  move validation before deduplication, or include those parameters in the cache key, so layers with different settings don't silently share the first mask

File: torchtitan/models/deepseek_v4/attention.py [hunk 7]
```diff
@@ -338,16 +355,73 @@ def forward(  # pyrefly: ignore[bad-param-name-override]
...
+            # [T, K] / [n_cmp, K]) so share_kv=True broadcasts them
+            # internally.
+            q_bhtk = q.transpose(0, 1).unsqueeze(0)
+            swa_k_b1tk = swa_k.unsqueeze(0).unsqueeze(0)
+            cmp_k_b1tk = cmp_k.unsqueeze(0).unsqueeze(0)
+
+            backend = "triton" if q.device.type == "cuda" else "eager"
+            out_bhtk = selected_attention(
+                q_bhtk,
+                swa_k_b1tk,
```
Comment (drisspg) [inline torchtitan/models/deepseek_v4/attention.py:416]: Does this transpose result in a query copy inside Attention Gym's `query.contiguous()`?  we shoudl do a path to make sure we swap contigousu to the min alignment needed

File: torchtitan/models/deepseek_v4/model.py [hunk 8]
```diff
@@ -175,8 +177,36 @@ def __init__(self, config: Config):
...
+
+            cache_key = (key, seqlen, device)
+            cached = self.mask_cache.get(cache_key)
+            if cached is None:
+                cached = inner.build_block_mask(seqlen=seqlen, device=device)
+                self.mask_cache[cache_key] = cached
```
Comment (drisspg) [inline torchtitan/models/deepseek_v4/model.py:207]: can we add cached-versus-uncached output and gradient comparisons for SWA/HCA, including cache reuse and a sequence-length change id ont think we are going through cache path in tests right?

File: torchtitan/models/deepseek_v4/attention.py (hunk 1 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py:325]: Hi Tianyu, what kind of perf comparison were you thinking of? In terms of end to end performance, the old model logs looked like this:
```
step:  6  loss:  4.07696  grad_norm:  2.3680  memory:  6.39GiB(3.47%)  tps: 11,274  tflops: 36.96  mfu: 1.48%
step:  7  loss:  3.94543  grad_norm:  1.7404  memory:  6.39GiB(3.47%)  tps: 11,293  tflops: 37.03  mfu: 1.48%
step:  8  loss:  3.84256  grad_norm:  1.9111  memory:  6.39GiB(3.47%)  tps: 11,209  tflops: 36.75  mfu: 1.47%
```
After after the changes, it's roughly 3x faster:
```
step:  6  loss:  4.07704  grad_norm:  2.3684  memory:  5.94GiB(3.22%)  tps: 33,094  tflops: 108.50  mfu: 4.34%
step:  7  loss:  3.94542  grad_norm:  1.7406  memory:  5.94GiB(3.22%)  tps: 32,667  tflops: 107.10  mfu: 4.28%
step:  8  loss:  3.84257  grad_norm:  1.9117  memory:  5.94GiB(3.22%)  tps: 32,430  tflops: 106.32  mfu: 4.25%
```

File: torchtitan/models/deepseek_v4/attention.py (hunk 2 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py]: Thanks for catching that, I fixed it.

File: torchtitan/models/deepseek_v4/attention.py (hunk 4 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py:361]: Thanks, that would be helpful.

File: torchtitan/models/deepseek_v4/attention.py (hunk 5 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py:418]: Currently, production configs aren't handled; I was thinking of waiting until the flash attention change got merged.

File: torchtitan/models/deepseek_v4/model.py (hunk 6 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/model.py:249]: Thanks for catching this, I fixed it

File: torchtitan/models/deepseek_v4/model.py (hunk 8 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/model.py:257]: Ok, I'll add this. We aren't going through the cache path for normal tests.

File: torchtitan/models/deepseek_v4/attention.py (hunk 3 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py:371]: That makes sense, I added a pr for this: https://github.com/meta-pytorch/attention-gym/pull/540

File: torchtitan/models/deepseek_v4/attention.py (hunk 5 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py:418]: Thanks for fixing up the flash attention change! I updated the code to route head_dim=512 to the cute backend.

File: torchtitan/models/deepseek_v4/attention.py (hunk 7 above)
Comment (floatingtrees) [reply torchtitan/models/deepseek_v4/attention.py:408]: I think this is only a problem for the triton backend; FA4 only cares about the last dimension being contiguous and aligned. Do you think this is a major issue if it only affects the debug models?

**Comment by floatingtrees** [comment]
I got it to use the newest version of attention gym, so the integration tests should pass now. I also fixed the lints that were failing due to some attention mask changes. Some lints still fail, but those are unrelated to this PR; they also fail on commit 1584a1833edeec22361f05114b643be2a64d4825 on main.

**Comment by drisspg** [comment]
doing 1 more pass; lets way till we publish 1 more relase for kda-cp branch before integrating this

File: torchtitan/models/deepseek_v4/attention.py [hunk 9]
```diff
@@ -338,16 +353,68 @@ def forward(  # pyrefly: ignore[bad-param-name-override]
...
+            q_bhtk = q.transpose(0, 1).unsqueeze(0)
+            swa_k_b1tk = swa_k.unsqueeze(0).unsqueeze(0)
+            cmp_k_b1tk = cmp_k.unsqueeze(0).unsqueeze(0)
+
+            backend = "eager"
+            if q.device.type == "cuda":
+                backend = (
+                    "cute"
+                    if head_dim == 512
+                    and torch.cuda.get_device_capability(q.device) == (10, 0)
```
Comment (drisspg) [inline torchtitan/models/deepseek_v4/attention.py:403]: We need https://github.com/Dao-AILab/flash-attention/pull/2882 to land for V4 Flash's 64 query heads and TP-sharded head counts. We should also cut a new Attention Gym release that relaxes the CuTe backend's hard-coded `h == 128` check and includes `scale`, then pin that release consistently in pyproject.toml, requirements.txt, and .ci/docker/requirements.txt. Landing FA4 alone isn't enough: both the current Gym pin and 499404b still reject non-128 head counts before calling FA4. Let's also ensure the installed flash-attn-4 version includes #2882.

**Comment by tianyu-l** [comment]
what's the status of this PR

**Comment by drisspg** [comment]
I'll take over @tianyu-l i was waiting to land my fix for smaller headda in fa4 but I'll do some routing to triton before that lands

**Comment by yinli-systems** [comment]
Hi — I took a pass at the current integration gap without opening a competing PR.

I ported this PR's intent onto current `main` (`0e4a2803b`) here: [yinli-systems/torchtitan@8214c85b2](https://github.com/yinli-systems/torchtitan/commit/8214c85b2).

The port addresses the drift since the last update:

- uses the `attn-gym==0.0.13` `gather_attn` API already pinned on `main`, including `scale`;
- leaves fused backend selection to Attention Gym's current CuTe/Triton routing, so it no longer depends on the closed [flash-attention#2882](https://github.com/Dao-AILab/flash-attention/pull/2882) path for non-128 or TP-sharded head counts;
- restores model-level SWA/HCA mask caching on the current `DSV4FlexInnerAttention` / model-config structure;
- constructs static SWA/HCA block metadata analytically, rather than materializing the cached implementation's full `[L, KV]` count and Boolean tensors;
- retains an uncached fallback for direct inner-attention callers.

Coverage includes cache reuse and sequence-length invalidation, incompatible-layer validation before deduplication, cached-vs-fresh output/gradient parity, CSA `BlockMask`-vs-`gather_attn` output/gradient parity, asymmetric block sizes, and a 32K assertion that `mask_mod` captures no token-pair tensor.

Current validation:

- CPU/reference: `5 passed`, plus `8` subtests.
- Focused formatting/lint/doc hooks pass; targeted Pyrefly 0.45.1 reports `0 errors`.
- A separate RTX 5090 benchmark of the same analytic static-mask builder at 32K reduced construction peak allocation from 8.21 GB to 3.21 MB and median construction time from 19.89 ms to 1.07 ms. These are mask-construction measurements, not end-to-end training speedups.
- The current-main CUDA output/gradient and cache suite is queued; I will edit this comment with the exact result rather than claim it early.

If this direction matches the takeover plan, please feel free to cherry-pick the commit or reuse the tests/analytic builder while reshaping the final backend routing.

AI disclosure: the implementation and wording were AI-assisted; I reviewed the code, ran the checks above, and kept only claims supported by the recorded evidence.
